servers/src/main/java/tachyon/master/TachyonMaster.java

```diff
@@ -203,168 +188,168 @@ public String getWebBindHost() {
-            } catch (TTransportException e) {
-              LOG.error(e.getMessage(), e);
-              throw Throwables.propagate(e);
-            }
-            mMasterMetricsSystem.registerSource(mMasterInfo.getMasterSource());
-            mMasterMetricsSystem.start();
-            mWebServer.addHandler(mMasterMetricsSystem.getServletHandler());
-            mWebServer.startWebServer();
-            LOG.info("The master (leader) server started @ " + mMasterAddress);
-            mMasterServiceServer.serve();
-            LOG.info("The master (previous leader) server ended @ " + mMasterAddress);
-            mJournal.close();
-          }
-        } else {
-          if (running) {
-            mMasterServiceServer.stop();
-            running = false;
-          }
-        }
-
-        CommonUtils.sleepMs(LOG, 100);
-      }
-    } else {
-      try {
-        setup();
-      } catch (IOException e) {
-        LOG.error(e.getMessage(), e);
-        throw Throwables.propagate(e);
-      } catch (TTransportException e) {
-        LOG.error(e.getMessage(), e);
-        throw Throwables.propagate(e);
-      }
+  protected void startServingWebServer() {
+    // start web ui
+    mWebServer = new MasterUIWebServer(ServiceType.MASTER_WEB, NetworkAddressUtils.getBindAddress(
+        ServiceType.MASTER_WEB, mTachyonConf), this, mTachyonConf);
+    mWebServer.startWebServer();
+  }
 
-      mMasterMetricsSystem.registerSource(mMasterInfo.getMasterSource());
-      mMasterMetricsSystem.start();
-      mWebServer.addHandler(mMasterMetricsSystem.getServletHandler());
-      mWebServer.startWebServer();
-      LOG.info("Tachyon Master version " + Version.VERSION + " started @ " + mMasterAddress);
-      mMasterServiceServer.serve();
-      LOG.info("Tachyon Master version " + Version.VERSION + " ended @ " + mMasterAddress);
+  protected void startServingRPCServer() {
+    // set up multiplexed thrift processors
+    TMultiplexedProcessor processor = new TMultiplexedProcessor();
+    processor.registerProcessor(mBlockMaster.getProcessorName(), mBlockMaster.getProcessor());
+    processor.registerProcessor(mFileSystemMaster.getProcessorName(),
+        mFileSystemMaster.getProcessor());
+    processor.registerProcessor(mRawTableMaster.getProcessorName(), mRawTableMaster.getProcessor());
+
+    // create master thrift service with the multiplexed processor.
+    mMasterServiceServer = new TThreadPoolServer(new TThreadPoolServer.Args(mTServerSocket)
+        .maxWorkerThreads(mMaxWorkerThreads).minWorkerThreads(mMinWorkerThreads)
+        .processor(processor).transportFactory(new TFramedTransport.Factory())
+        .protocolFactory(new TBinaryProtocol.Factory(true, true)));
+
+    // start thrift rpc server
+    mIsServing = true;
+    mStartTimeMs = System.currentTimeMillis();
+    mMasterServiceServer.serve();
+  }
+
+  protected void stopServing() throws Exception {
+    if (mMasterServiceServer != null) {
+      mMasterServiceServer.stop();
```

Comment:
Since we create a new MasterServiceServer when we start, shall we set it to null after stop?